### ISSUE TITLE
D100 with —add-noqa breaks shebang lines

### ISSUE THREAD

**Issue by pjspereira** [astral-sh/ruff#14442]
The `—add-noqa` breaks shebang lines with the D100 rule on files with no module documentation. Can the shebang line be ignored and the D100 error be assigned to the second line, if it exists?

Example output from `ruff check --add-noqa --select D100 --isolated .`:

```python
#!/usr/bin/env python  # noqa: D100

import sys

if __name__==‘__main__’:
    print(sys.argv)
```

If this script is marked executable, the noqa comment will be passed as a filename to python, which will crash the interpreter, since the file does not exist.

Keywords: shebang D100
Ruff version: 0.6.4

**Comment by MichaReiser** [comment]
That sounds reasonable. We need to change the diagnostic location to after the first line

**Comment by MichaReiser** [comment]
Yeah, it's certainly not ideal and thanks for the heads up for the RyeCharm integration. 

We could consider detecting this in the fix loop but my worry is that it won't be possible to determine if the edit should be applied before or after the shebang line (e.g. what if an edit changes the shebang line). But maybe we can? By only applying said logic if its an insertion at offset 0 and a shebang line is present?

**Comment by MusicalNinjaDad** [comment]
Does that mean that all file-level rules should:
- expect their file-level `# noqa:` to be directly below the shebang, or would the convention be to add a blank line in-between?
- apply their `--fix`es to the correct place?

I imagine this would be best done in a central location - e.g. where the noqa directives / fixes are parsed? using something like the EXE rules do: `if let Some(shebang) = ShebangDirective::try_extract(comment)`

**Comment by MichaReiser** [comment]
I'm not quiet sure what the right fix is. There's no clear good solution. The problem is that our noqa system expects that the suppression comments are on the same line as the diagnostic's range. 

We could explore if the following works:

* In `add_noqa`: If we detect that we're on the first line and the first line is a shebang, add the suppression right after the shebang line
* When finding suppression comments and the diagnostic range starts at `0:0` and there is a shebang, respect suppressions on the next line.

**Comment by MusicalNinjaDad** [comment]
I'd be happy to take a look at this, if it's not urgent and you're ok with that. Although I don't want to step on @AlexWaygood 's toes...

**Comment by MichaReiser** [comment]
Feel free to take a look. I don't think @AlexWaygood is working on this. Just a heads up. We're currently very busy with the type checker and reviews could be somewhat delayed because of that.

**Comment by AlexWaygood** [comment]
I think this is one of the issues I assigned myself to at the beginning of the year because it was listed as a possible starter task for @ntBre (I didn't want anybody else taking it in case it was something he was interested in working on 😄). Both @ntBre and I are now somewhat busy with other tasks, so anybody who wants to look at this should feel free to do so!

**Comment by getehen** [comment]
Hi, I would like to take that, no one working on it ?
